Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm

```diff
@@ -138,36 +155,243 @@ - (void)webViewDidClose:(WKWebView *)webView
 
 @end
 
+#if PLATFORM(IOS_FAMILY)
+static void* kvoContext = &kvoContext;
+#endif
+
 @interface _WKWebExtensionActionWebView : WKWebView
+
+@property (nonatomic, readonly) BOOL contentSizeHasStabilized;
+
 @end
 
 @implementation _WKWebExtensionActionWebView {
-    WeakPtr<WebKit::WebExtensionAction> _webExtensionAction;
+    WeakPtr<WebExtensionAction> _webExtensionAction;
+    CGSize _previousContentSize;
 }
 
-- (instancetype)initWithFrame:(CGRect)frame configuration:(WKWebViewConfiguration *)configuration webExtensionAction:(WebKit::WebExtensionAction&)action
+- (instancetype)initWithFrame:(CGRect)frame configuration:(WKWebViewConfiguration *)configuration webExtensionAction:(WebExtensionAction&)action
 {
     if (!(self = [super initWithFrame:frame configuration:configuration]))
         return nil;
 
     _webExtensionAction = action;
 
+#if PLATFORM(IOS_FAMILY)
+    [self.scrollView addObserver:self forKeyPath:@"contentSize" options:NSKeyValueObservingOptionNew context:kvoContext];
+#endif
```

Comment:
I think this could maybe use a comment explaining why this is only on iOS. Say for Mac we use `invalidateIntrinsicContentSize` to listen for size changes of the WKWebView.